File: csharp/1052.grumpy-bookstore-owner.cs
/*
 * @lc app=leetcode id=1052 lang=csharp
 *
 * [1052] Grumpy Bookstore Owner
 */

// @lc code=start
public class Solution {
    public int MaxSatisfied(int[] customers, int[] grumpy, int X) {
        int n = customers.Length, ret = 0, acc = 0, macc = 0;
        for (int i = 0; i < n; ++i) {
            if (grumpy[i] == 0) {
                ret += customers[i];
            } else {
                acc += customers[i];
            }
            if (i >= X && grumpy[i - X] == 1) {
                acc -= customers[i - X];
            }
            if (acc > macc) {
                macc = acc;
            }
        }
        return ret + macc;
    }
}
// @lc code=end

File: java/1052.grumpy-bookstore-owner.java
/*
 * @lc app=leetcode id=1052 lang=java
 *
 * [1052] Grumpy Bookstore Owner
 */

// @lc code=start
class Solution {
    public int maxSatisfied(int[] customers, int[] grumpy, int minutes) {
        int h = 0, n = customers.length, l = 0, g = 0, m = 0;
        for (int r = 0; r < n; ++r) {
            if (grumpy[r] == 0) {
                h += customers[r];
            } else {
                g += customers[r];
            }
            if (r - l + 1 > minutes) {
                if (grumpy[l] == 1) {
                    g -= customers[l];
                }
                ++l;
            }
            m = Math.max(m, g);
        }
        return m + h;
    }
}
// @lc code=end

